hudi-aws/src/main/java/org/apache/hudi/aws/transaction/lock/S3StorageLockClient.java

```diff
@@ -0,0 +1,252 @@
+    //    The update operation will include an ifMatch(expectedEtag) condition, meaning the update will only
+    //    succeed if the current file's ETag exactly matches expectedEtag.
+    //    If the actual ETag of the file on S3 differs from expectedEtag, the update attempt will fail.
+    // - If expectedEtag is null:
+    //    We assume that the file does not currently exist on S3.
+    //    The operation will use ifNoneMatch("*"), which instructs S3 to create the file only if it doesn't already exist.
+    //    If a file with the same name is present (i.e., there is an existing ETag), the creation attempt will fail.
+    if (expectedEtag == null) {
+      putRequestBuilder.ifNoneMatch("*");
+    } else {
+      putRequestBuilder.ifMatch(expectedEtag);
+    }
+
+    PutObjectResponse response = s3Client.putObject(putRequestBuilder.build(), RequestBody.fromBytes(bytes));
+    String newEtag = response.eTag();
+
+    return new StorageLockFile(lockData, newEtag);
+  }
+
+  private LockUpsertResult handleUpsertS3Exception(S3Exception e) {
+    int status = e.statusCode();
+    if (status == PRECONDITION_FAILURE_ERROR_CODE) {
+      logger.warn("OwnerId: {}, Lockfile modified by another process: {}", ownerId, lockFilePath);
+      return LockUpsertResult.ACQUIRED_BY_OTHERS;
+    } else if (status == CONDITIONAL_REQUEST_CONFLICT_ERROR_CODE) {
+      logger.warn("OwnerId: {}, Retriable conditional request conflict error: {}", ownerId, lockFilePath);
+    } else if (status == RATE_LIMIT_ERROR_CODE) {
+      logger.warn("OwnerId: {}, Rate limit exceeded for: {}", ownerId, lockFilePath);
+    } else if (status >= INTERNAL_SERVER_ERROR_CODE_MIN) {
+      logger.warn("OwnerId: {}, internal server error for: {}", ownerId, lockFilePath, e);
+    } else {
+      logger.error("OwnerId: {}, Error writing lock file: {}", ownerId, lockFilePath, e);
+    }
+
+    return LockUpsertResult.UNKNOWN_ERROR;
+  }
+
+  private static Functions.Function2<String, Properties, S3Client> createDefaultS3Client() {
+    return (bucketName, props) -> {
+      Region region;
+      boolean requiredFallbackRegion = false;
+      try {
+        region = DefaultAwsRegionProviderChain.builder().build().getRegion();
+      } catch (SdkClientException e) {
+        // Fallback to us-east-1 if no region is found
+        region = Region.US_EAST_1;
+        requiredFallbackRegion = true;
+      }
+
+      S3Client s3Client = S3Client.builder().credentialsProvider(
+              HoodieAWSCredentialsProviderFactory.getAwsCredentialsProvider(props)).region(region).build();
+      if (requiredFallbackRegion) {
+        GetBucketLocationResponse bucketLocationResponse = s3Client.getBucketLocation(
+                GetBucketLocationRequest.builder().bucket(bucketName).build());
+        String regionString = bucketLocationResponse.locationConstraintAsString();
+        if (!StringUtils.isNullOrEmpty(regionString)) {
```

Comment:
can we also check against default region. what incase bucket is actually in Region.US_EAST_1. we do not need to reinstantiate right?

Reply:
In fact if bucket is in us east 1 the location constraint is null. I will add a comment as such.